cdap-docs/admin-manual/source/operations/logging.rst

```diff
@@ -50,28 +102,409 @@ CDAP supports logging through standard `SLF4J (Simple Logging Facade for Java)
-14 days. Changes can be made to ``logback-container.xml``, and the ``cdap-master`` process will need to be restarted
-for the modified logback to take effect.
+**Custom logging pipelines** are configured by a combination of the settings in the 
+:ref:`cdap-site.xml <appendix-cdap-site.xml>` file and a "logback" file used
+to specify the custom pipeline. The XML file is placed in the
+``log.process.pipeline.config.dir``, a local directory on the CDAP Master node that is scanned
+for log processing pipeline configurations. Each pipeline is defined by a file in the
+Logback XML format, with ``.xml`` as the file name extension.
+
+These :ref:`properties <appendix-cdap-default-logging>` control the CDAP log pipeline:
+
+.. generated by code in conf.py for this manual
+.. include:: /../target/_includes/logging-pipeline-configuration-1.rst
+
+.. _custom-log-pipeline-logging-configuration:
+
+These :ref:`properties <appendix-cdap-default-logging>` control both the CDAP log pipeline
+and custom log pipelines:
+
+.. generated by code in conf.py for this manual
+.. include:: /../target/_includes/logging-pipeline-configuration-2.rst
+
+The ``log.process.pipeline.*`` properties can be over-ridden and specified at the
+custom pipeline level by providing a value in a pipeline's "logback" file for any of
+these properties.
+
+
+Logging Framework
+=================
+
+.. highlight:: console
+
+This diagram shows in greater detail the components and steps CDAP follows when logging
+programs of an application and system services with the logging framework:
+
+  - Logs are collected from individual programs running in YARN containers.
+
+  - YARN containers have their own local log files.
+
+  - CDAP System Services run (depending on the service) either on cluster edge nodes or in
+    YARN containers. Where they run determines the file that configures that service's
+    logging.
+
+  - CDAP logs are published to Kafka topics.
+
+  - The Log Saver Service (*Log.saver*) is configured to read log messages for the
+    ``logs.user-v2`` Kafka topic. The number of log saver instances can be scaled to process
+    the Kafka partitions in parallel, if needed. Log.saver reads the messages from Kafka,
+    buffers and sorts them in-memory, before persisting them to HDFS. These messages are
+    written to files corresponding to the programs the log messages belong to.
+
+    *Note:* These files are configured to rotate based on time and size; they can be
+    changed using the properties ``log.pipeline.cdap.file.max.size.bytes`` and
+    ``log.pipeline.cdap.file.max.lifetime.ms`` in the :ref:`cdap-site.xml
+    <appendix-cdap-default-logging>` file as described in 
+    :ref:`logging-monitoring-logging-pipeline-configuration`.
+
+  - Custom pipelines can be configured by adding an XML file in a prescribed location.
+    Each pipeline buffers log messages in-memory and sorts them based on their timestamp. 
```

Comment:
Perhaps structure this as follows:
- Log.saver, by default, runs only the CDAP log pipeline. This performs:
  - read, group, sort, persist to files
  - emit metrics about each program/log level
- In addition, custom log pipelines can be configured...